### DISCUSSION TITLE
Cannot load more than two pages (browser tabs) at the site time. v 2.2 on IIS

### DISCUSSION THREAD

**Discussion by horacioj** [oqtane/oqtane.framework#1633] [Q&A]
Hi,

- Standard Oqtane (version 2.2) IIS (win 10) installation based on https://www.oqtane.org/Resources/Blog/PostId/542/installing-oqtane-on-iis
- The singe update I made was to allow all verbs, in the web.config, I added the `modules` node.
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
    <modules runAllManagedModulesForAllRequests="true">
        <remove name="WebDAVModule"/>
    </modules>	
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
      <aspNetCore processPath="dotnet" arguments=".\Oqtane.Server.dll" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" hostingModel="inprocess" />
    </system.webServer>
  </location>
</configuration>
```
- I accepted all the defaults, including the default site creation
- The site was created as some.localtest.me (the IIS binding, etc. is fine)
- If I open a browser tab on the home page, everything is fine
- I open a 2nd browser tab (or a different browser, the problem is the same) for some.localtest.me/private, everything is fine
- I try to open a 3rd browser tab (or a different browser, the problem is the same) for some.localtest.me/mypage, it never loads up. The GET for /mypage stays forever as '(pending)',
- If I close one of the two already opened pages (tabs in the same or different browsers), the /mypage is finally loaded. Note: not even trying to browse to some.localtest.me/_framework/blazor.server.js is loaded. It also stays pending while the problem is happening.

- The problem also happens when having many sites (some.localtest.me, some2.localtest.me, some3.localtest.me). With any combination of sites, pages, and browsers, the 3rd one gets stuck until one of the others two is closed.

This looks like a problem (sessions, maybe?) on IIS, but I never faced it before.
Or could it be a problem in Oqtane?

Thanks

EDIT:
This looks similar (https://stackoverflow.com/questions/63693352/blazor-server-of-connections-per-use-is-limited-by-the-browser), but the WebSocket Protocol is already installed.

EDIT 2:
The limit for Win 10 is at least 10 connections, it looks like Oqtane is making a lot of connections (SignalR), which quickly reach the limit when just trying to open a third page.

**Comment by sbwalker** [comment]
@horacioj  I have no idea why you are experiencing issues. I am guessing it has something to do with the specific configuration of your own environment. https://www.oqtane.net is running Oqtane 2.2 on Blazor Server on IIS. https://www.dnfprojects.com is also running Oqtane 2.2 on Blazor Server on IIS. They are not exhibiting the behavior you are describing.

**Reply by horacioj** [reply]
@sbwalker, IIS on windows server allows a lot more connections.

IIS on Windows 10 (especially, the Home edition) allows only 10 or 20 simultaneous connections.

Other Blazor servers apps on my Win 10 machine doesn't show this problem. Only Oqtane is allowing me only two pages open at the same time, which are consuming all the allowed connections. 

10 (or 20) connections for only two pages seems to be showing something is wrong.

I'll try to see how can I monitor the actual connections to verify the limit and measure how many connections each Oqtane page is consuming.

**Reply by sbwalker** [reply]
Sorry I did not realize you were taking about a development environment. Yes, there are limitations on IIS on Windows 10 Pro/Home for the number of connections. This limits the scalability but is generally not an issue for general development activities. It's possible that other Blazor Server applications may be able to support more tabs on IIS as they are probably not as comprehensive in terms of their functionality... however since they are still based on SignalR they will hit the limit if you try to connect too many clients.

**Reply by horacioj** [reply]
I'm commenting this in case it isn't the the normal and expected behavior. Is it?

I tested again other Blazor server (not Oqtane) apps and each page takes 2 connections (some times 3).
An "empty" Oqtane page (standard theme, logged-in (*), and with a single HTML module with just one 'lorm ipsum' paragraph takes 10 connections, opening a 2nd page with similar content takes 7 more connections, bringing a total of 17 connections for two basically empty pages.

(*) not logged-in shows the same results 10+7

A "hard reload" of both pages shows 24 connections. After some minutes (leaving both pages unattended) this drops to 7 connections. However, opening a 3rd page doesn't work. So, I don't know if the Performance Monitor is not showing the proper data or what... If I close one of the two opened pages, the 3rd page continues loading fine. After some (many) minutes, Perfmon shows 20 connections.

![oqtane-connections](https://user-images.githubusercontent.com/984821/131389336-cb0034c2-ad91-4083-83b3-a671b08c4572.jpg)

![oqtane-connections-II](https://user-images.githubusercontent.com/984821/131389360-8653144d-cdb2-4648-8532-d8fd92513e4a.jpg)

hard reload of both pages
![oqtane-connections-III-hard_reload](https://user-images.githubusercontent.com/984821/131389370-4c83f61a-da05-47e7-92be-407bfa4d62af.jpg)

Some time later
![oqtane-connections-IV](https://user-images.githubusercontent.com/984821/131389387-5031e82d-95f2-44bc-a24c-2f979206dc28.jpg)

**Reply by sbwalker** [reply]
@horacioj I am not sure what it is that you are trying to investigate in your development environment - as obviously Blazor Server behaves much differently in a production environment. Were you planning on creating an application that required many different browser tabs to be open in your development environment?

**Reply by sbwalker** [reply]
It is not clear to me what you are measuring in terms of "connections". Are these connections between the browser and server? If so there are different types of connections to consider. There is the SignalR connection between the browser tab and the server for UI rendering and events. And then there are HttpClient connections ( ie. AJAX requests ) between the browser and server for retrieving data. Depending on the functionality of a specific page there may be many HttpClient connections required.

**Reply by sbwalker** [reply]
More info on this from the Microsoft docs:

IIS limitations on Windows client OS

Windows 10 and Windows 8.x are client operating systems. IIS on client operating systems has a limit of 10 concurrent connections. SignalR's connections are:

    Transient and frequently re-established.
    Not disposed immediately when no longer used.

The preceding conditions make it likely to hit the 10 connection limit on a client OS. When a client OS is used for development, we recommend:

    Avoid IIS.
    Use Kestrel or IIS Express as deployment targets.

**Reply by horacioj** [reply]
Thank you, @sbwalker. I was worried about a possible problem in production (IIS / Oqtane in server mode) when many (+2k) users could hitting the site at the same time.

**Reply by sbwalker** [reply]
This Microsoft blog contains some information about the scalability of Blazor Server applications in production:

https://devblogs.microsoft.com/aspnet/blazor-server-in-net-core-3-0-scenarios-and-performance/

"A single Standard_D1_v2 instance on Azure (1 vCPU, 3.5 GB memory) could handle over 5,000 concurrent users without any degradation in latency. A Standard_D3_V2 instance (4 vCPU, 14GB memory) handled well over 20,000 concurrent clients."

**Reply by horacioj** [reply]
I was worried about Oqtane (not Blazor itself) making too many connections. But now I don't think this is a problem Thanks!
